sdv/metadata/single_table.py

```diff
@@ -536,65 +550,125 @@ def _determine_sdtype_for_objects(self, data):
-                    'key.'
-                )
-
-        candidates = []
-        first_pii_field = None
-        for column, column_meta in self.columns.items():
-            sdtype = column_meta['sdtype']
-            column_data = data[column]
+        try:
+            clean_data = column_data.dropna()
+            dtype = clean_data.infer_objects().dtype.kind
+            pk_candidate = False
             has_nan = column_data.isna().any()
             valid_potential_primary_key = column_data.is_unique and not has_nan
-            sdtype_in_reference = sdtype in self._REFERENCE_TO_SDTYPE.values()
-            if sdtype == 'id':
-                candidates.append(column)
-                if len(candidates) > 1:
-                    self.columns[column]['sdtype'] = 'unknown'
-                    self.columns[column]['pii'] = True
-
-            elif sdtype_in_reference and first_pii_field is None and valid_potential_primary_key:
-                first_pii_field = column
-
-        data.columns = original_columns
-        if candidates:
-            return candidates[0]
-        if first_pii_field:
-            return first_pii_field
+            sdtype = self._detect_pii_column(column_name) or self._detect_id_column(column_name)
+
+            if sdtype is None:
+                if dtype in self._DTYPES_TO_SDTYPES:
+                    sdtype = self._DTYPES_TO_SDTYPES[dtype]
+                elif dtype in self._NUMERICAL_DTYPES:
+                    sdtype, pk_candidate = self._determine_sdtype_for_numbers(
+                        column_data, valid_potential_primary_key
+                    )
 
-        return None
+                elif dtype == 'O':
+                    sdtype, pk_candidate = self._determine_sdtype_for_objects(
+                        column_data, valid_potential_primary_key
+                    )
+
+                if sdtype is None:
+                    table_str = f"table '{table_name}' " if table_name else ''
+                    error_message = (
+                        f"Unsupported data type for {table_str}column '{column_name}' "
+                        f"(kind: {dtype}). The valid data types are: 'object', "
+                        "'int', 'float', 'datetime', 'bool'."
+                    )
+                    raise InvalidMetadataError(error_message)
+            elif valid_potential_primary_key:
+                if sdtype == 'id':
+                    pk_candidate = True
+                else:
+                    pii_pk_candidates.append(column_name)
+
+            if pk_candidate:
+                pk_candidates.append(column_name)
+
+            return sdtype, dtype
+
+        except InvalidMetadataError as e:
+            raise e
+        except Exception as e:
+            self._handle_detection_error(e, column_name, table_name)
+
+    def _select_primary_key(self, infer_sdtypes, pk_candidates, pii_pk_candidates):
+        """Select the primary from a list of candidates.
```

Comment:
Select the primary from a list of candidates.
```suggestion
        """Select the primary key from a list of candidates.
```